### PR TITLE
Guzzle middleware

### PR DESCRIPTION
**Pull request by iamluc**
Continuation of https://github.com/Tolerance/Tolerance/pull/86

### CODE REVIEW

**Comment by csarrazi** [comment]
See my comment in csarrazi/CsaGuzzleBundle#191

Why not completely integrate the middleware, and thus simplify the code? I don't really see the point of using semantic configuration only for letting the user replace a service with his own (this can be done using a service definition). Also, the middleware are automatically registered should the user create his own middleware with a DI tag.

The only thing needed would be to provide documentation explaining how to integrate the middleware with CsaGuzzleBundle. :)

File: src/Tolerance/Bridge/Guzzle/RetryMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public function __invoke(RequestInterface $request, array $options)
+    {
+        $nextHandler = $this->nextHandler;
+        $operation = new PromiseOperation(function () use ($nextHandler, $request, $options) {
+            return $nextHandler($request, $options);
+        });
+
+        $waiterFactory = $this->waiterFactory;
+        $runner = new RetryPromiseOperationRunner($waiterFactory($options), $this->errorVoter);
```
Comment (iamluc) [inline src/Tolerance/Bridge/Guzzle/RetryMiddleware.php]: Should be better to get the runner as a dependency injection.

But the problem is that we need a new waiter each time as many requests could run simultaneously and we don't want to reset the state of the waiter.

File: src/Tolerance/Bridge/Symfony/Bundle/ToleranceBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -30,6 +30,7 @@ public function getConfigTreeBuilder()
             ->append($this->getOperationRunnersNode())
             ->append($this->getMetricsNode())
             ->booleanNode('operation_runner_listener')->defaultTrue()->end()
+            ->booleanNode('guzzle')->defaultFalse()->end()
```
Comment (iamluc) [inline src/Tolerance/Bridge/Symfony/Bundle/ToleranceBundle/DependencyInjection/Configuration.php]: Could be removed and replaced by a new section in the documentation, but provide a simple "plug'n'play" feature.

File: src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php [hunk 3]
```diff
@@ -95,6 +100,11 @@ protected function onTerminate(PromiseOperation $operation, ThrowableCatcherVote
             try {
                 $this->waitStrategy->wait();
             } catch (WaiterException $waiterException) {
+                // If it is a Guzzle Promise, use Guzzle's objects.
+                if ($promise instanceof PromiseInterface) {
+                    return $fulfilled ? new FulfilledPromise($value) : new RejectedPromise($value);
```
Comment (iamluc) [inline src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php]: This is needed to not break existing code.
Without this change, caller will catch a Tolerance's `PromiseException`. But with the change, the caller will catch a Guzzle's exception (`ClientException`/`ServerException`/`ConnectException`)

**Comment by iamluc** [comment]
ping @sroze 
Will you have time to review this soon ?

**Comment by iamluc** [comment]
Documentation added !

**Comment by sroze** [comment]
Fabulous, thank you very much @iamluc! 👍

**Comment by tyx** [comment]
Cannot wait for the next release :dancing_men:

**Pull request merged**